### PR TITLE
Add groth16 verification for zkid transaction signatures

### PR DESCRIPTION
**Pull request by heliuchuan**
### Description

This adds the ability to verify groth16 proofs as an option to authenticate transactions.  This builds on #11681 and implements the privacy-preserving path based on Groth16 ZKPs.

### `SignedGroth16Zkp`

```rust

pub struct SignedGroth16Zkp {
    pub proof: Groth16Zkp,
    /// The signature of the proof signed by the private key of the `ephemeral_pubkey`.
    pub non_malleability_signature: EphemeralSignature,
}

pub struct Groth16Zkp {
    a: G1Bytes,
    b: G2Bytes,
    c: G1Bytes,
}

pub struct G1Bytes(pub(crate) [u8; G1_PROJECTIVE_COMPRESSED_NUM_BYTES]);

pub struct G2Bytes(
    #[serde(with = "BigArray")] pub(crate) [u8; G2_PROJECTIVE_COMPRESSED_NUM_BYTES],
);
```

#### Signature size

`Groth16Zkp.a` -> 32 bytes
`Groth16Zkp.b` -> 64 bytes
`Groth16Zkp.c` -> 32 bytes
Groth16Zkp = ~ 128 bytes

`SignedGroth16Zkp.proof` -> 128 bytes
`SignedGroth16Zkp.non_malleability_signature` -> 66 bytes (when Ed25519 is used, schemes in the future could differ)
SignedGroth16Zkp = ~ 194 bytes

`ZkIdSignature.sig` -> 1 bytes (enum) + ~ 194 bytes
`ZkIdSignature.jwt_header` ->  ~ 100 bytes
`ZkIdSignature.exp_timestamp_secs` -> 8 bytes (u64)
`ZkIdSignature.ephemeral_pubkey` -> 34 bytes (when Ed25519 is used but up to 93 bytes)
`ZkIdSignature.ephemeral_signature` -> 66 bytes (when Ed25519 is used, schemes in the future could differ)
ZkIdSignature= ~ 403 bytes

`ZkIdPublicKey.iss` -> ~27 bytes (google)
`ZkIdPublicKey.idc` -> ~32 byte
ZkIdPublicKey= ~ 59 bytes

**Total Size Estimate**
Total for ZkId Groth16 Authenticator = ~462

### Signature verification
To verify an SignedGroth16Zkp with a ZkIdSignature, the following checks must pass

1. The SignedGroth16Zkp.non_malleability_signature must be verified with the ZkIdSignature.ephemeral_pubkey to be the valid signature for SignedGroth16Zkp.proof
2. Check that `ZkIdSignature.exp_timestamp_secs` < `now` (the current time on-chain at validation time)
3. Fetch the JWK associated with `ZkIdSignature.jwt_header` and construct the public_inputs_hash from the JWK, ZkIdSignature, and ZkIdPublicKey.  Verify that SignedGroth16Zkp.proof is a valid proof with the public_inputs_hash and corresponding verification key.

Todos after circuit finalization
- [ ] Update claim max length constants when finalized
- [ ] add recovery field

### Test Plan
Smoke tested and unit tested

### CODE REVIEW

File: testsuite/smoke-test/src/zkid.rs [hunk 1]
```diff
@@ -18,90 +18,14 @@ use aptos_types::{
         authenticator::{AnyPublicKey, EphemeralPublicKey, EphemeralSignature},
         SignedTransaction,
     },
-    zkid::{IdCommitment, OpenIdSig, Pepper, ZkIdPublicKey, ZkIdSignature, ZkpOrOpenIdSig},
+    zkid::{IdCommitment, OpenIdSig, Pepper, ZkIdPublicKey, ZkIdSignature, ZkpOrOpenIdSig, Groth16Zkp},
 };
 use move_core_types::account_address::AccountAddress;
 use std::time::Duration;
 
-async fn get_latest_jwkset(rest_client: &Client) -> PatchedJWKs {
```
Comment (heliuchuan) [inline testsuite/smoke-test/src/zkid.rs:26]: just moving this to the end of the file

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 2]
```diff
@@ -80,32 +120,40 @@ pub fn validate_zkid_authenticators(
...
+                        let public_inputs_hash =
+                            get_public_inputs_hash(zkid_sig, zkid_pub_key, &rsa_jwk).map_err(
+                                |_| {
+                                    VMStatus::error(
+                                        StatusCode::INVALID_SIGNATURE,
+                                        Some("Could not compute public inputs hash".to_owned()),
+                                    )
+                                },
+                            )?;
+                        let chain_id = get_chain_id(resolver)?;
```
[the remark this answers is not in the document]
Comment (heliuchuan) [reply aptos-move/aptos-vm/src/zkid_validation.rs:139]: how would i do that? where is it stored?

[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/zkid_validation.rs]: It is currently under `move_vm_ext/vm.rs` inside `MoveVmExt` struct. I guess you would need to write a getter to access it from AptosVM (which stores `MoveVmExt` under the hood)

File: types/src/circom.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    PreparedVerifyingKey::from(vk)
+}
+
+fn parse_field_element(s: &str) -> Result<Fq, CryptoMaterialError> {
+    s.parse::<Fq>()
+        .map_err(|_e| CryptoMaterialError::DeserializationError)
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct G1 {
```
Comment (mstraka100) [inline types/src/circom.rs:90]: G1Projective instead of G1?

File: types/src/circom.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    let mut frs = poseidon_bn254::pad_and_pack_bytes_to_scalars_with_len(
+        sig.ephemeral_pubkey.to_bytes().as_slice(),
+        MAX_EPK_BYTES,
+    )?;
+
+    // Add the id_commitment as a scalar
+    frs.push(poseidon_bn254::pack_bytes_to_one_scalar(&pk.idc.0)?);
+
+    let iat_val = 1700255944;
+    // Add the exp_timestamp_secs as a scalar
```
Comment (mstraka100) [inline types/src/circom.rs:183]: typo?

File: types/src/jwks/rsa.rs [hunk 5]
```diff
@@ -33,6 +37,19 @@ impl RSA_JWK {
         }
     }
 
+    pub fn to_poseidon_scalar(&self) -> Result<ark_bn254::Fr> {
+        let mut modulus = base64::decode_config(&self.n, URL_SAFE_NO_PAD)?;
+        modulus.reverse(); // This is done to match the circuit TODO(zkid): finalize the jwk hashing scheme.
```
Comment (mstraka100) [inline types/src/jwks/rsa.rs:42]: "This is done to match the circuit, which requires the modulus in a verify specific format due to how RSA verification is implemented"

**Review by mstraka100 — APPROVED** [review]
Some minor comments, LGTM overall. Should probably be reviewed by someone more involved with the JWK consensus work

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 6]
```diff
@@ -38,11 +40,41 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
+}
+
 pub fn validate_zkid_authenticators(
     transaction: &SignedTransaction,
     resolver: &impl AptosMoveResolver,
     _session: &mut SessionExt,
     _log_context: &AdapterLogSchema,
+    chain_id: ChainId,
 ) -> anyhow::Result<(), VMStatus> {
     // TODO(ZkIdGroth16Zkp): The ZKP/OpenID sig verification does not charge gas. So, we could have DoS attacks.
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/zkid_validation.rs:79]: cc @vgao1996 

We probably should not write such comments in public repo?😂 At least say that "we should charge" :)

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 7]
```diff
@@ -38,11 +40,41 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
+        .get_jwk(&zkid_pub_key.iss, &jwt_header_parsed.kid)
+        .map_err(|_| {
+            VMStatus::error(
+                StatusCode::INVALID_SIGNATURE,
+                Some("JWK not found".to_owned()),
+            )
+        })?;
+
+    let jwk = JWK::try_from(jwk_move_struct).map_err(|_| {
+        VMStatus::error(
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/zkid_validation.rs:64]: Seems like 
```
VMStatus::error(
            StatusCode::INVALID_SIGNATURE,
            Some("Some message".to_owned()),
        )
```
is used a lot, maybe use a macro?

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 8]
```diff
@@ -38,11 +40,41 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
+    Ok(jwk)
+}
+
 pub fn validate_zkid_authenticators(
     transaction: &SignedTransaction,
     resolver: &impl AptosMoveResolver,
     _session: &mut SessionExt,
     _log_context: &AdapterLogSchema,
+    chain_id: ChainId,
 ) -> anyhow::Result<(), VMStatus> {
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/zkid_validation.rs:78]: Why this is anyhow? You can use a normal Result?

File: aptos-move/aptos-vm/src/zkid_validation.rs (hunk 6 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/zkid_validation.rs:79]: Actually, you cannot even charge because this is a prologue?

Comment (vgao1996) [reply aptos-move/aptos-vm/src/zkid_validation.rs:79]: This has not be enabled, right? If so then we are fine now.

Good to be cautious though!

Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/zkid_validation.rs:79]: But it will be?😅

Comment (vgao1996) [reply aptos-move/aptos-vm/src/zkid_validation.rs:79]: Yeah just make it clear that proper gas metering needs to be implemented before we could expose this to our users.

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 9]
```diff
@@ -38,79 +46,70 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
     Ok(jwks)
 }
 
+fn get_jwk_for_zk_authenticator(
```
Comment (alinush) [inline aptos-move/aptos-vm/src/zkid_validation.rs:49]: Note to self: How come we didn't have this function before? Maybe it was specialized for the OpenIdSig path.

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 10]
```diff
@@ -38,79 +46,70 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
+                    proof
+                        .verify_proof(public_inputs_hash, chain_id)
+                        .map_err(|_| invalid_signature!("Proof verification failed"))?;
+                },
+                JWK::Unsupported(_) => return Err(invalid_signature!("JWK is not supported")),
+            },
             ZkpOrOpenIdSig::OpenIdSig(openid_sig) => {
                 match jwk {
                     JWK::RSA(rsa_jwk) => {
-                        // TODO(OpenIdSig): Implement batch verification for all RSA signatures in
```
Comment (alinush) [inline aptos-move/aptos-vm/src/zkid_validation.rs:112]: @heliuchuan not sure why you removed this TODO. Can I add it back?

File: types/src/circom.rs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    PreparedVerifyingKey::from(vk)
+}
+
+fn parse_field_element(s: &str) -> Result<Fq, CryptoMaterialError> {
+    s.parse::<Fq>()
+        .map_err(|_e| CryptoMaterialError::DeserializationError)
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct G1Projective {
```
Comment (alinush) [inline types/src/circom.rs:91]: This format is not efficient. We need to store G1 and G2 bytes as compressed bytes.

Let's discuss asynchronously.

File: types/src/circom.rs [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    // Add the epk as padded and packed scalars
+    let mut frs = poseidon_bn254::pad_and_pack_bytes_to_scalars_with_len(
+        sig.ephemeral_pubkey.to_bytes().as_slice(),
+        MAX_EPK_BYTES,
+    )?;
+
+    // Add the id_commitment as a scalar
+    frs.push(ark_bn254::Fr::from_le_bytes_mod_order(&pk.idc.0));
+
+    let iat_val = 1700255944;
```
Comment (alinush) [inline types/src/circom.rs:183]: We cannot merge until this is taken care of right?

File: types/src/jwks/rsa.rs [hunk 13]
```diff
@@ -33,6 +37,20 @@ impl RSA_JWK {
         }
     }
 
+    pub fn to_poseidon_scalar(&self) -> Result<ark_bn254::Fr> {
```
Comment (alinush) [inline types/src/jwks/rsa.rs:40]: I think this should be part of the `poseidon_bn254` module, where all hashing is defined. So other applications can use it easily (e.g., proving service).

File: types/src/zkid.rs [hunk 14]
```diff
@@ -32,12 +36,12 @@ pub const IDC_NUM_BYTES: usize = 32;
...
 // TODO(ZkIdGroth16Zkp): determine what our circuit will accept
 
-pub const MAX_EPK_BYTES: usize = 93; // Supports public key lengths of up to 93 bytes.
-pub const MAX_ISS_BYTES: usize = 248;
-pub const MAX_AUD_VAL_BYTES: usize = 248;
-pub const MAX_UID_KEY_BYTES: usize = 248;
-pub const MAX_UID_VAL_BYTES: usize = 248;
-pub const MAX_JWT_HEADER_BYTES: usize = 248;
+pub const MAX_EPK_BYTES: usize = 3 * poseidon_bn254::BYTES_PACKED_PER_SCALAR; // Supports public key lengths of up to 93 bytes.
+pub const MAX_ISS_BYTES: usize = 5 * poseidon_bn254::BYTES_PACKED_PER_SCALAR;
```
Comment (alinush) [inline types/src/zkid.rs:40]: Why is this specified as 5 * BYTES_PACKED_PER_SCALAR rather than directly?  For `MAX_EPK_BYTES` I understand, but for this one, not sure.

File: types/src/zkid.rs [hunk 15]
```diff
@@ -225,13 +229,45 @@ impl Claims {
     }
 }
 
-pub type G1 = Vec<String>;
-pub type G2 = Vec<Vec<String>>;
-#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+#[derive(
+    Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize, CryptoHasher, BCSCryptoHash,
+)]
 pub struct Groth16Zkp {
```
Comment (alinush) [inline types/src/zkid.rs:235]: Note to self: We'll need to ensure that arkworks deserialization checks subgroup membership properly.

File: types/src/zkid.rs [hunk 16]
```diff
@@ -225,13 +229,45 @@ impl Claims {
...
+    a: G1Projective,
+    b: G2Projective,
+    c: G1Projective,
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct SignedGroth16Zkp {
+    pub proof: Groth16Zkp,
+    /// The signature of the proof signed by the private key of the `ephemeral_pubkey`.
+    pub proof_signature: EphemeralSignature,
```
Comment (alinush) [inline types/src/zkid.rs:245]: Is this the non-malleability signature or the training wheels signature?

Comment (heliuchuan) [reply types/src/zkid.rs:245]: non-malleability signature

File: types/src/zkid.rs (hunk 14 above)
Comment (heliuchuan) [reply types/src/zkid.rs:40]: the reason is we'd ultimate never choose a value that is not a multiple of BYTES_PACKED_PER_SCALAR right?

File: types/src/zkid.rs (hunk 16 above)
Comment (alinush) [reply types/src/zkid.rs]: Okay. I renamed.

File: types/src/jwks/rsa.rs [hunk 17]
```diff
@@ -33,6 +37,20 @@ impl RSA_JWK {
...
+        let mut modulus = base64::decode_config(&self.n, URL_SAFE_NO_PAD)?;
+        modulus.reverse(); // This is done to match the circuit, which requires the modulus in a verify specific format due to how RSA verification is implemented
+                           // TODO(zkid): finalize the jwk hashing scheme.
+        let mut scalars = modulus
+            .chunks(24) // Pack 3 64 bit limbs per scalar, so chunk into 24 bytes per scalar
+            .map(|chunk| {
+                poseidon_bn254::pack_bytes_to_one_scalar(chunk).expect("chunk converts to scalar")
+            })
+            .collect::<Vec<ark_bn254::Fr>>();
+        scalars.push(ark_bn254::Fr::from(RSA_MODULUS_BYTES as i32));
```
Comment (zjma) [inline types/src/jwks/rsa.rs:50]: `RSA_MODULUS_BYTES` is 256. What if self is an RS384 key?
Currently if a provider returns a RSA384 key, jwk consensus can support it.

File: types/src/zkid.rs (hunk 15 above)
Comment (zjma) [reply types/src/zkid.rs:235]: iirc they check by default and has those `*_unchecked` apis.

File: types/src/circom.rs [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+    PreparedVerifyingKey::from(vk)
+}
+
+fn parse_field_element(s: &str) -> Result<Fq, CryptoMaterialError> {
+    s.parse::<Fq>()
+        .map_err(|_e| CryptoMaterialError::DeserializationError)
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct G1Projective {
```
[the remark this answers is not in the document]
Comment (zjma) [reply types/src/circom.rs:91]: same across this file, and the file name itself

File: aptos-move/aptos-vm/src/zkid_validation.rs (hunk 6 above)
Comment (alinush) [reply aptos-move/aptos-vm/src/zkid_validation.rs]: Will take care of this.

File: types/src/jwks/rsa.rs (hunk 17 above)
Comment (alinush) [reply types/src/jwks/rsa.rs:50]: JWK consensus should definitely support it, for the zkless path to work, even though the ZK path will not support it.

@heliuchuan, you are best familiar with the code, so we just need to make sure that the ZK path properly rejects Groth16-based zkID signatures whose JWK is **not** an RSA 256-byte key.

File: types/src/circom.rs (hunk 12 above)
Comment (heliuchuan) [reply types/src/circom.rs:183]: Done

File: types/src/jwks/rsa.rs (hunk 13 above)
Comment (heliuchuan) [reply types/src/jwks/rsa.rs:39]: i will refactor this later, I want to move the whole of jwks in aptos crypto.  just moving this function right now is a little annoying without the whole struct.  I think we want to do that anyway.  Left todo.

File: types/src/jwks/rsa.rs (hunk 17 above)
Comment (heliuchuan) [reply types/src/jwks/rsa.rs:49]: Done, added check for 256 length key.

File: types/src/zkid.rs (hunk 15 above)
Comment (heliuchuan) [reply types/src/zkid.rs:235]: confirmed

File: types/src/circom.rs (hunk 11 above)
Comment (heliuchuan) [reply types/src/circom.rs:102]: Updated to take 32 and 64 byte representations of G1 and G2 projective respectively.

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 19]
```diff
@@ -38,74 +46,67 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
+    let jwk_move_struct = jwks
+        .get_jwk(&zkid_pub_key.iss, &jwt_header_parsed.kid)
+        .map_err(|_| invalid_signature!("JWK not found"))?;
+
+    let jwk =
+        JWK::try_from(jwk_move_struct).map_err(|_| invalid_signature!("Could not parse JWK"))?;
+    Ok(jwk)
+}
+
 pub fn validate_zkid_authenticators(
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/zkid_validation.rs:66]: So we also have https://github.com/aptos-labs/aptos-core/blob/b40329261475357cc4527e65668f198aa2787fc9/aptos-move/aptos-vm/src/aptos_vm.rs#L2248, but it seems nothing is verified there and just flags are checked. Is this intentional?

Comment (heliuchuan) [reply aptos-move/aptos-vm/src/zkid_validation.rs:66]: Yes. We do all our flag checking there.  I agree it's not ideal.

Here's [the thread](https://github.com/aptos-labs/aptos-core/pull/11681#discussion_r1458192227) about this from the other pr

tldr - it's an annoying refactor.  We'd need to update validate_signed_transaction to take in Vec<SingleKeyAuthenticator> or move all the feature flag checking into validate_signed_transaction.

down to do it if you have a preference

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 20]
```diff
@@ -1385,8 +1385,15 @@ impl AptosVM {
                 None,
             ));
         }
+        let chain_id = self.move_vm.get_chain_id();
 
-        zkid_validation::validate_zkid_authenticators(transaction, resolver, session, log_context)?;
+        zkid_validation::validate_zkid_authenticators(
```
[the remark this answers is not in the document]
Comment (heliuchuan) [reply aptos-move/aptos-vm/src/aptos_vm.rs:1390]: Makes sense to me - done.

File: aptos-move/aptos-vm/src/zkid_validation.rs [hunk 21]
```diff
@@ -38,74 +46,67 @@ fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> anyhow::Result<Patched
...
     transaction: &SignedTransaction,
     resolver: &impl AptosMoveResolver,
     _session: &mut SessionExt,
```
[the remark this answers is not in the document]
Comment (heliuchuan) [reply aptos-move/aptos-vm/src/zkid_validation.rs:69]: Was going to but can be removed now

[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/zkid_validation.rs]: we can keep it in this case

File: aptos-move/aptos-vm/src/zkid_validation.rs (hunk 19 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/zkid_validation.rs:65]: yes, refactoring is annoying. Let's add a TODO for now so we do not forget, and revisit as a part of more substantial VM refactoring

File: types/src/bn254_circom.rs [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+        buf
+    }};
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct Bn254G1ProjectiveCompressed(pub(crate) [u8; G1_PROJECTIVE_COMPRESSED_NUM_BYTES]);
+
+impl Bn254G1ProjectiveCompressed {
+    pub fn new(x: &str, y: &str) -> anyhow::Result<Self> {
+        let g1 = ark_bn254::G1Projective::new_unchecked(
```
Comment (alinush) [inline types/src/bn254_circom.rs:125]: Note to self: Using `_unchecked` sounds like a problem.

File: types/src/bn254_circom.rs [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+
+    fn try_into(self) -> Result<ark_bn254::G1Affine, CryptoMaterialError> {
+        let g1_projective: ark_bn254::G1Projective = self.try_into()?;
+        Ok(g1_projective.into())
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct Bn254G2ProjectiveCompressed(
+    #[serde(with = "BigArray")] pub(crate) [u8; G2_PROJECTIVE_COMPRESSED_NUM_BYTES],
```
Comment (alinush) [inline types/src/bn254_circom.rs:177]: Why do you need this?

File: types/src/jwks/rsa.rs [hunk 24]
```diff
@@ -37,8 +37,13 @@ impl RSA_JWK {
         }
     }
 
+    // TODO(zkid): Move this to aptos-crypto so other services can use this
     pub fn to_poseidon_scalar(&self) -> Result<ark_bn254::Fr> {
         let mut modulus = base64::decode_config(&self.n, URL_SAFE_NO_PAD)?;
+        // The circuit only supports RSA256
+        if modulus.len() != RSA_MODULUS_BYTES {
```
Comment (alinush) [inline types/src/jwks/rsa.rs:44]: Oops, I also added this check after fetching the JWK. I think it is best done there rather than in the hashing method?

File: types/src/bn254_circom.rs (hunk 22 above)
Comment (heliuchuan) [reply types/src/bn254_circom.rs:125]: it does, but checked uses assert! and can potentially panic.  it's annoying.  We could call the checks after (would need to convert to affine).

But we need to deserialize to affine eventually anyway which has the checks.  but yeah not ideal

File: types/src/jwks/rsa.rs (hunk 24 above)
Comment (heliuchuan) [reply types/src/jwks/rsa.rs:44]: Personally I want to put it in the hashing method, we already have very circuit specific logic in the hashing method.

**Review by alinush — APPROVED** [review]
